### PR TITLE
[RW-3650][risk=no] Add label support to our core CheckBox class and add type-safety and unit tests.

### PR DESCRIPTION
**Pull request by gjuggler**
I ended up digging into the checkbox issue a bit more on Friday. I think this is the most balanced approach to fixing the issue. The crux of the matter is this: should a checkbox manage its own checked state, or should it always blindly render the value passed in via props?

After taking stock of existing usage in the Workbench, I realized that in most scenarios, it's fine to allow checkboxes manage their own state. Wrapping components can pass in the initial value, and listen for changes by providing an .onChange(checked) prop.

In one or two use cases, though, we seem to be expecting the checkbox to always render the value passed in via props. The wrapping component implements the toggling behavior by listening to onChange events, updating its own state, and using that state variable to pass into the CheckBox's 'props'. This pattern should be disfavored, since it causes wrapping components to effectively re-implement the checkbox toggling logic.

I also took the opportunity to add type-safe props and state to the CheckBox component, and a couple unit tests. It's a bit scary that, across our whole codebase, our core UI components have the least type-safety and unit testing. This makes it hard to safely improve or refactor our UI over time, so I'd like to (very incrementally) add these over time.

---
**PR checklist**

- [X] This PR meets the Acceptance Criteria in the JIRA story
- [X] The JIRA story has been moved to Dev Review
- [X] This PR includes appropriate unit tests
- [X] I have run and tested this change locally

### CODE REVIEW

File: ui/src/app/components/inputs.tsx [hunk 1]
```diff
@@ -119,64 +133,87 @@ export const RadioButton = ({ onChange, ...props }) => {
...
+  id?: string;
+  // An optional label to show alongside the input. Can be a plain string or
+  // any React node.
+  label?: React.ReactNode;
+  // Styles to apply to the label element. Only relevant when label is non-null.
+  labelStyle?: React.CSSProperties;
+  // Indicates whether the CheckBox should be responsible for managing its own
+  // state. When false, the HTML input will always be rendered with the value of
+  // the passed 'checked' prop.
+  manageOwnState: boolean;
```
Comment (als364) [inline ui/src/app/components/inputs.tsx]: Should we file a ticket to deprecate this?

File: ui/src/app/components/inputs.tsx [hunk 2]
```diff
@@ -119,64 +133,87 @@ export const RadioButton = ({ onChange, ...props }) => {
...
+    const {
+      checked, disabled, label, labelStyle, onChange, manageOwnState, style, wrapperStyle,
+      ...otherProps
+    } = this.props;
+    const uniqueId = fp.uniqueId('checkbox');
+
+    const maybeDisabledOverrides = disabled ? styles.disabledStyle : {};
+
+    const input = <input
+      id={uniqueId}
```
Comment (als364) [inline ui/src/app/components/inputs.tsx]: not `props.id ? props.id : uniqueId`?

File: ui/src/app/components/tables.tsx [hunk 3]
```diff
@@ -20,14 +20,14 @@ export const TwoColPaddedTable = ({style = {}, header = false, headerLeft = '',
   headerRight = '', cellWidth = {left: '50%', right: '50%'}, contentLeft, contentRight}) => {
   return <FlexColumn style={{...style}}>
     {header &&
-      <FlexRow style={{height: '100%'}}>
-        <PaddedTableCell left={true} leftWidth={cellWidth.left}
+      <FlexRow key='header' style={{height: '100%'}}>
```
Comment (als364) [inline ui/src/app/components/tables.tsx]: this seems unrelated. Did this miss a linter somewhere?

File: ui/src/app/pages/analysis/confirm-playground-mode-modal.tsx [hunk 4]
```diff
@@ -27,10 +27,9 @@ export class ConfirmPlaygroundModeModal extends React.Component<Props, State> {
     };
   }
 
-  toggleChecked() {
-    const newState = !this.state.checked;
-    this.setState({checked: newState});
-    Cookies.set(ConfirmPlaygroundModeModal.DO_NOT_SHOW_AGAIN, String(newState));
+  toggleChecked(checked: boolean) {
+    this.setState({checked: checked});
```
Comment (als364) [inline ui/src/app/pages/analysis/confirm-playground-mode-modal.tsx]: shouldn't this be `checked: !checked`?

Comment (als364) [reply ui/src/app/pages/analysis/confirm-playground-mode-modal.tsx]: or does this happen after the actual checked state changes? if that's the case we should change the name of this function since it is no longer toggling checked

File: ui/src/app/pages/login/account-creation/account-creation-survey.tsx [hunk 5]
```diff
@@ -110,6 +102,7 @@ export class AccountCreationSurvey extends React.Component<AccountCreationSurvey
     // Toggle Includes removes the element if it already exist and adds if not
     const attributeList = toggleIncludes(value, this.state.profile.demographicSurvey[attribute]);
     this.setState(fp.set(['profile', 'demographicSurvey', attribute], attributeList));
+    console.log(this.state);
```
Comment (als364) [inline ui/src/app/pages/login/account-creation/account-creation-survey.tsx]: extraneous

File: ui/src/app/pages/workspace/workspace-edit.tsx [hunk 6]
```diff
@@ -521,7 +532,7 @@ export const WorkspaceEdit = fp.flow(withRouteConfigData(), withCurrentWorkspace
     }
 
     specificPopulationCheckboxSelected(populationEnum): boolean {
-      return fp.includes(populationEnum, this.state.workspace.researchPurpose.populationDetails);
+      return !!fp.includes(populationEnum, this.state.workspace.researchPurpose.populationDetails);
```
Comment (als364) [inline ui/src/app/pages/workspace/workspace-edit.tsx]: uh oh, what probably hilarious bug made the boolifier necessary?

File: ui/src/app/pages/workspace/workspace-edit.tsx [hunk 7]
```diff
@@ -707,13 +718,14 @@ export const WorkspaceEdit = fp.flow(withRouteConfigData(), withCurrentWorkspace
...
+          >
           <CheckBox
-                 style={{height: '.66667rem', marginRight: '.31667rem', marginTop: '1.2rem'}}
-          onChange={v => this.setState({cloneUserRole: v})}/>
-          <WorkspaceEditSection header='Copy Original workspace Collaborators'
-            description='Share cloned workspace with same collaborators'/>
-        </FlexRow>
+            style={styles.checkboxStyle}
+            label='Share workspace with the same set of collaborators'
+            labelStyle={styles.text}
```
Comment (als364) [inline ui/src/app/pages/workspace/workspace-edit.tsx]: this results in the same style as the original styling? even with no margins set?

File: ui/src/app/pages/workspace/workspace-edit.tsx [hunk 8]
```diff
@@ -707,13 +718,14 @@ export const WorkspaceEdit = fp.flow(withRouteConfigData(), withCurrentWorkspace
...
+        <WorkspaceEditSection header='Options for duplicate workspace'
+          >
           <CheckBox
-                 style={{height: '.66667rem', marginRight: '.31667rem', marginTop: '1.2rem'}}
-          onChange={v => this.setState({cloneUserRole: v})}/>
-          <WorkspaceEditSection header='Copy Original workspace Collaborators'
-            description='Share cloned workspace with same collaborators'/>
-        </FlexRow>
+            style={styles.checkboxStyle}
+            label='Share workspace with the same set of collaborators'
```
Comment (als364) [inline ui/src/app/pages/workspace/workspace-edit.tsx]: This is better wording, thank you

File: ui/src/app/pages/workspace/workspace-edit.tsx [hunk 9]
```diff
@@ -744,6 +757,7 @@ export const WorkspaceEdit = fp.flow(withRouteConfigData(), withCurrentWorkspace
                 .map((rp, i) =>
                   <WorkspaceCategory shortDescription={rp.shortDescription}
                     longDescription={rp.longDescription} key={i}
+                                     uniqueId={rp.uniqueId}
```
Comment (als364) [inline ui/src/app/pages/workspace/workspace-edit.tsx]: indentation is off here.

File: ui/src/app/pages/workspace/workspace-edit.tsx [hunk 10]
```diff
@@ -734,6 +746,7 @@ export const WorkspaceEdit = fp.flow(withRouteConfigData(), withCurrentWorkspace
                 .map((rp, i) =>
                   <WorkspaceCategory shortDescription={rp.shortDescription} key={i}
                     longDescription={rp.longDescription}
+                     uniqueId={rp.uniqueId}
```
Comment (als364) [inline ui/src/app/pages/workspace/workspace-edit.tsx]: indentation is off here.

File: ui/src/app/components/inputs.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+import {CheckBox} from './inputs';
+
+function findInput(wrapper: (ShallowWrapper|ReactWrapper)): (ShallowWrapper|ReactWrapper) {
+  return wrapper.find('input[type="checkbox"]').first();
+}
+
+function clickCheckbox(wrapper: (ShallowWrapper|ReactWrapper)) {
+  const currentChecked = findInput(wrapper).prop('checked');
+  findInput(wrapper).simulate('change', {target: {checked: !currentChecked}});
```
[the remark this answers is not in the document]
Comment (als364) [reply ui/src/app/components/inputs.spec.tsx]: Boooo

This code feels like we're just testing that 1=1.

File: ui/src/app/pages/workspace/workspace-edit.tsx (hunk 7 above)
Comment (als364) [reply ui/src/app/pages/workspace/workspace-edit.tsx]: I do like the appearance of this better than the old appearance, including the larger checkbox.

File: ui/src/app/components/inputs.spec.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (gjuggler) [reply ui/src/app/components/inputs.spec.tsx]: Yeah... I'm not thrilled about it, but (1) it demonstrates expected usage and does test a modicum of the state-management flow, and (2) stupid tests are still better than null tests, esp. in a core component library like this :)

I'll probably poke at this a bit more in the future... stuff like this bugs me, as I pretty strongly believe UI stuff should be tested as realistically as possible, e.g. simulating true clicks on the DOM. But that does seem to go somewhat against the grain of typical React unit test patterns. Maybe the real answer is, we have defense in depth and this kind of stuff will get caught at an end-to-end test stage.

**Comment by gjuggler** [comment]
Hi @dolbeew – I'd like to get your eyes on these changes, in case I'm missing something in how the checkboxes are working in the sidebar-component. And also more generally as a heads up re: the change to how checkboxes work by default.
